Config to require engines to make a ship fly/float with balloons

EnderBoy9217 opened the pull request:
A simple change that adds a config option that requires flying ships to have an active engine.

The option is disabled by default to fit the existing Eureka system.

This change is meant for survival worlds where flying ships are currently just undeniably better than making water ships. It is essentially a balancing change intended for multiplayer survival worlds. I have already distributed some custom JARs to some server owners who were interested in this.

It does this by simply refusing to calculate the force applied by balloons unless an engine is currently providing power.

[Video Demonstration](https://cdn.discordapp.com/attachments/1323005693405565040/1323005694839881739/Minecraft_2024.12.29_-_14.03.08.02.mp4?ex=6772f091&is=67719f11&hm=e92e591434b8c70d4a042d4b70c3ce41e0c13a4e2a98fba6f7e1ef4e1309f38f&)

EnderBoy9217 committed:
Require Engines to Fly (with config)

EnderBoy9217 committed:
Add config

millennIumAMbiguity commented:
Engine heat should contribute to a buffer/multiplier.
E.g. balloonMultiplier = min(1, balloonMultiplier + engineHeat * heatSpeedCofig / sqrt(totalBalloons));

EnderBoy9217 committed:
Add new configs for heat scaling

EnderBoy9217 committed:
Update to scale with engines

EnderBoy9217 commented:
I don't believe there is a need for a buffer, as the engine already provides heat and extra force for a while after the last item is used.

However, I did add a multiplier based on the heat and number of engines, which is also configurable.

millennIumAMbiguity commented:
you need to scale the extraForceLinear. 
```kt
var unscaledExtraForce = (extraForceLinear - EurekaConfig.SERVER.enginePowerLinearMin) / EurekaConfig.SERVER.enginePowerLinear
```
sadly we don't actually know how many total engines we have, but we can calculate how hot a singular engine would be.
you also don't need to calculate the force twice. `balloonForceProvided` gets multiplied by getPlayerUpwardVel

I suggest just using this:
```kt
val balloonForceProvided = balloons * forcePerBalloon + balloons * EurekaConfig.SERVER.extraForcePerHotBalloon * clamp(EurekaConfig.SERVER.hotBalloonHeatMultiplier * (extraForceLinear - EurekaConfig.SERVER.enginePowerLinearMin) / EurekaConfig.SERVER.enginePowerLinear, 0, 1)
```
This way, it's easy to mix between how difficult the ballon heat should be.
`extraForcePerHotBalloon` would be the amount of lift a fully heated balloon adds.
`hotBalloonHeatMultiplier` can be seen as how many engines at full heat it requires to fully heat all balloons.

EnderBoy9217 commented:
Apologies for the *VERY* delayed response. However, the goal of this PR, and what many users expressed interest in, was that the ships would by default be in a falling state (balloon force not applying) unless engines were running, not just increasing balloon force depending on the engine force. Your suggestion would provide the same amount of balloon force when engines aren't running, and also wouldn't scale properly with changes to the ForcePerBalloon (which can be changed by either the MassPerBalloon Config or anything affecting gravity), which would lead to near-negligible change. And while the `hotBalloonHeatMultiplier` would be related the number of engines required to achieve maximum extra thrust, it is an inverse relationship that would be confusing for a user to change.

However, I'll admit that my own implementation also had some problems. I noticed in testing that basically turning on any engine would give most ships the ability to fly, and if it didn't the problem could simply be fixed by adding another balloon, basically either large ships hardly required any engines or small ships too many, which is confusing an unintuitive.

I have come up with a new solution, and have created this graph with the current potential solution, your suggested solution, and a new algorithm I will soon update this PR with. This solution requires the number of engines enabled to be scaled with the number of balloons and as such the mass, as this PR is specifically to help balance survival mode, this would bring a much-needed addition rather than simply having a single engine running. This would also allow for all extra config related to this algorithm to be condensed into one extra config `Max Balloons Per Engine`, which is easy for users to change and understand.

I have modeled all three algorithms and their respective configs changes in [this graph](https://www.desmos.com/calculator/wirbxmsdyg). EnginePowerLinear and ForcePerBalloon are using normalized values so the entire graph is easier to see and navigate.

I will update this PR soon and hopefully have time to test it tonight.

EnderBoy9217 commented:
Also
> you also don't need to calculate the force twice. balloonForceProvided gets multiplied by getPlayerUpwardVel
@millennIumAMbiguity 


I'm assuming this means that this algorithm does not need to apply to the `balloonForceProvided` variable in the `getPlayerUpwardVel` function, as it gets clamped later. I am unable to test this right now so please correct me if I misunderstood that.

EnderBoy9217 committed:
Merge branch 'ValkyrienSkies:1.18.x/main' into EnderBoy9217-engine-required

EnderBoy9217 committed:
Update Config to have one value

EnderBoy9217 commented:
The 5 default in the config is just an arbitrary untested number that would probably work, if you test this let me know how well that translates in-game and if it should be adjusted.

EnderBoy9217 committed:
Remove Balloons Per Engine Algorithm from getPlayerUpwardVel function

EnderBoy9217 committed:
Update Balloons Per Engine Algorithm

* Condenses config for the algorithm into one easier-to-use value
* Balloon power scales correctly with engine power
* Added a second divide by 0 case

EnderBoy9217 commented:
PR is confirmed working on fabric and correctly scales.

EnderBoy9217 committed:
Update .gitignore

EnderBoy9217 committed:
Clarify variable

millennIumAMbiguity committed:
Clean-up

 - Reverted gitignare. should be a separate issue pr with valid argument.
 - Reverted variable name "balloonForceProvided"
 - Removed the "flightRequiresEngine" config as setting maxBalloonsPerEngine to 0 gives same effect.
 - And other cleanups and
